File: rust/e4.rs
/*
Find the largest palindrome made from the product of two 3-digit numbers.
*/
use std::cmp;

fn euler_4() -> usize {
    let (mut candidate, mut etadidnac, mut max_palindrome) = (0, 0, 0);
    for i in 100..1000 {
        for j in 100..1000 {
            candidate = i*j;
            etadidnac = reverse(candidate);
            if candidate == etadidnac {
                max_palindrome = cmp::max(max_palindrome, candidate);
            }
        }
    }
    max_palindrome
}

fn reverse(mut num: usize) -> usize {
    let (mut digit, mut mun) = (0, 0);
    while num > 0 {
        digit = num % 10;
        mun *= 10;
        mun += digit;
        num = num / 10;
    }
    mun
}

fn main() {
    println!("{}", euler_4());
}

File: julia/e4.jl
#=
Find the largest palindrome made from the product of two 3-digit numbers.
=#

function euler_4()::Int
    #verified
    big_palindrome = 0
    for (i, j) in Base.Iterators.product(100:999,100:999)
        candidate = (i*j)
        #we're doing it the hard way, damnit - no tostring() method!
        etadidnac = reverse(candidate)
        if (etadidnac == candidate)
            big_palindrome = max(big_palindrome, candidate)
        end
    end
    big_palindrome
end

function reverse(num::Int)::Int
    i = 0
    mun = 0
    while num > 0
        digit = num % 10
        mun *= 10
        mun += digit
        num ÷= 10
        i += 1
    end
    mun
end

println(euler_4())
